### PR TITLE
Support .NET Standard 2.0

### PR DESCRIPTION
**Pull request by Bouke**
I recently came across this library, and it has a few features over MoreLinq that I could use: mostly the async versions. However for my use-case I'd need support for `netstandard2.0`, which is currently not being targeted (#153).

This PR is an attempt to add support for `netstandard2.0`. The build works, but there are a few test failures that I haven't looked into. I'm posting this PR mostly to get some feedback on the direction I've taken, and to see test the waters before I pour additional effort into this.

I've opted to use the `NETCOREAPP` constant to cover both `netstandard2.0` and `net48`. The test projects need to target `net48` to test the `netstandard2.0` version of the library.

One concern is that I've marked the polyfills as public, so that `Index` and `Range` are usable. It seems to work, but might be something to evaluate. The alternative is removing all those API for the `netstandard2.0` target.

### CODE REVIEW

File: Tests/SuperLinq.Async.Test/CountByTest.cs [hunk 1]
```diff
@@ -83,9 +83,11 @@ public async Task CountByWithSomeNullKeys()
...
-				KeyValuePair.Create((string?)null, 4),
-				KeyValuePair.Create((string?)"bar", 2),
-				KeyValuePair.Create((string?)"baz", 2));
+				CreatePair((string?)"foo", 2),
+				CreatePair((string?)null, 4),
+				CreatePair((string?)"bar", 2),
+				CreatePair((string?)"baz", 2));
 	}
+
+	private KeyValuePair<TKey, TValue> CreatePair<TKey, TValue>(TKey key, TValue value) => new(key, value);
```
Comment (viceroypenguin) [inline Tests/SuperLinq.Async.Test/CountByTest.cs:92]:
```suggestion
	private static KeyValuePair<TKey, TValue> CreatePair<TKey, TValue>(TKey key, TValue value) => new(key, value);
```

**Comment by viceroypenguin** [comment]
Thanks for the PR! I appreciate the interest, and I've been wanting to have this conversation for a long time, obviously. 

The biggest question I have is what to do about `Index` and `Range`. Obviously if they are `public`, then they will conflict with any other public definitions in other libraries that the consumer may already own. However, if we remove the `Index` and `Range` apis, then that reduces the potential offering for consumers.

Have you given any thought to the conflict that having two libraries with public `Index` and `Range` structs would raise?

File: Source/SuperLinq/Future.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+		actualValue = set.ElementAt(index);
+		return true;
+	}
+
+	public static HashSet<TSource> ToHashSet<TSource>(
+		this IEnumerable<TSource> source,
+		IEqualityComparer<TSource>? comparer)
+	{
+		if (source == null)
```
Comment (viceroypenguin) [inline Source/SuperLinq/Future.cs:41]: Unnecessary line, since `ArgumentNullException` will do this check.

**Comment by viceroypenguin** [comment]
Based on https://github.com/dotnet/runtime/pull/104170, I'd like to wait until `Microsoft.Bcl.Memory` is published to Nuget. Then we can depend directly on this package in ns2.0 instead of relying on Polysharp. This will remove the need to have a conversation around public access to `Index` and `Range`.

**Comment by Bouke** [comment]
Shipping `Index` / `Range` to consumers will not be a great experience, even if they also use PolySharp. I already had the "great" experience of resolving conflicts as both SuperLinq and SuperLinq.Async shipped these. So I see a few options going forward:

- Wait for https://github.com/dotnet/runtime/pull/104170 to ship: unknown time frame
- Ship PolySharp polyfills public: problematic
- Hide API using `Index` / `Range`, unhide once https://github.com/dotnet/runtime/pull/104170 ships

I've updated the PR to implement the last option: hide API using `Index` / `Range`. This enables going further with this PR and not be blocked by future updates to dependencies.

**Comment by viceroypenguin** [comment]
I'd rather see the `Index` and `Range` APIs excluded entirely rather than set to `internal`, I think?

File: Source/SuperLinq.Async/SuperLinq.Async.csproj [hunk 3]
```diff
@@ -124,6 +124,14 @@
     <ProjectReference Include="..\SuperLinq\SuperLinq.csproj" />
   </ItemGroup>
 
+	<ItemGroup Condition="$(TargetFramework) == 'netstandard2.0'">
+		<PackageReference Include="Microsoft.Bcl.HashCode" />
+		<PackageReference Include="System.Memory" PrivateAssets="All" />
```
Comment (viceroypenguin) [inline Source/SuperLinq.Async/SuperLinq.Async.csproj:129]:
```suggestion
		<PackageReference Include="System.Memory" />
```

File: Source/SuperLinq/SuperLinq.csproj [hunk 4]
```diff
@@ -161,6 +161,18 @@
     <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" PrivateAssets="All" />
     <PackageReference Include="Microsoft.SourceLink.GitHub" PrivateAssets="All" />
     <PackageReference Include="MinVer" PrivateAssets="All" />
+    <PackageReference Include="System.Memory" PrivateAssets="All" Condition="$(TargetFramework) == 'netstandard2.0'" />
```
Comment (viceroypenguin) [inline Source/SuperLinq/SuperLinq.csproj:164]:
```suggestion
```

File: Source/SuperLinq/SuperLinq.csproj [hunk 5]
```diff
@@ -161,6 +161,18 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition="$(TargetFramework) == 'netstandard2.0'">
+    <PackageReference Include="Microsoft.Bcl.HashCode" />
+    <PackageReference Include="System.Memory" PrivateAssets="All" />
```
Comment (viceroypenguin) [inline Source/SuperLinq/SuperLinq.csproj:169]:
```suggestion
    <PackageReference Include="System.Memory" />
```

**Comment by Bouke** [comment]
I've reconstruct Unshipped from netcoreapp3.1, removing Index/Range API. Sadly Rider doesn't support applying _all_ Roslyn fixits, and `dotnet format analyzers --diagnostics=RS0016` didn't produce entries for generated code.

**Comment by Bouke** [comment]
The test projects must not target netstandard2.0 (is not something you can "run"), but net48. Suggestions on how I replace `netstandard2.0` with `net48` in the test project's `TargetFramework`s?

```
<TargetFrameworks>$(TargetFrameworks.Replace("netstandard2.0","net48"))</TargetFrameworks>
```

This produces:

> Error: /usr/share/dotnet/sdk/9.0.100-preview.6.24328.19/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(100,5): error NETSDK1046: The TargetFramework value 'net48;netcoreapp3.1;net6.0;net8.0;net9.0' is not valid. To multi-target, use the 'TargetFrameworks' property instead. [/home/runner/work/SuperLinq/SuperLinq/Tests/SuperLinq.Async.Test/SuperLinq.Async.Test.csproj::TargetFramework=net48 netcoreapp3.1 net6.0 net8.0 net9.0]

**Comment by Bouke** [comment]
Do you have a preference over how to resolve the remaining issues? I.e. the conditional `#if ..` makes the resulting indentation incorrect here:

```cs
#if !NETSTANDARD
		if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
		{
#endif
			// Clear the elements so that the gc can reclaim the references
			Array.Clear(_nodes, 0, Count);
			_elementIndex.Clear();
#if !NETSTANDARD
		}
#endif
```

File: Source/SuperLinq/Collections/UpdatablePriorityQueue.cs [hunk 6]
```diff
@@ -650,12 +653,16 @@ public void EnqueueRangeMinimum(IEnumerable<TElement> elements, TPriority priori
 	/// </summary>
 	public void Clear()
 	{
+#if !NETSTANDARD
```
Comment (Bouke) [inline Source/SuperLinq/Collections/UpdatablePriorityQueue.cs:656]: Do you have a preference over how to address the code style issues here? The `#if` makes the resulting indentation incorrect.

**Comment by viceroypenguin** [comment]
@Bouke It builds now at least! I'm still gonna work through what I think is best regarding `Index`/`Range` though.

**Comment by viceroypenguin** [comment]
Ok, take a look and tell me what you think. With `NO_INDEX` as the define, it should be easy to restore these APIs later when it's possible. These are separate from the `NETCOREAPP` define, which will remain due to BCL differences.

**Comment by Bouke** [comment]
> Ok, take a look and tell me what you think. With `NO_INDEX` as the define, it should be easy to restore these APIs later when it's possible. These are separate from the `NETCOREAPP` define, which will remain due to BCL differences.

Distinguishing between `NO_INDEX` and `NETCOREAPP` makes the code easier to follow, that's a win. Whether to disable the additional APIs as you did is up to you; it'll be easy to add them in the future also. So overall LGTM!

**Comment by Bouke** [comment]
Maybe you can set my expectations on when you might be creating a new release with these changes included?

**Comment by viceroypenguin** [comment]
v6.2.0 is out now! 😄

**Pull request merged**